File: backend/human_delta.py
"""Human Delta scoring (parity with frontend lib/humanDelta.js)."""

from token_estimate import estimate_tokens_by_model


def reduction_percent(before: float, after: float) -> float:
    if before <= 0:
        return 0.0
    raw = ((before - after) / before) * 100
    return max(0.0, min(100.0, round(raw, 1)))


def compute_human_delta(before_tokens: float, after_tokens: float) -> dict:
    efficiency_score = reduction_percent(before_tokens, after_tokens)
    if efficiency_score > 60:
        impact = "LOW IMPACT"
    elif efficiency_score > 30:
        impact = "MEDIUM IMPACT"
    else:
        impact = "HIGH IMPACT"
    return {
        "beforeTokens": before_tokens,
        "afterTokens": after_tokens,
        "efficiencyScore": efficiency_score,
        "impactLevel": impact,
    }


def human_delta_from_prompts(raw: str, optimized: str, model: str = "GPT-4") -> dict:
    b = estimate_tokens_by_model(raw, model)
    a = estimate_tokens_by_model(optimized, model)
    return compute_human_delta(b, a)

File: frontend/lib/humanDelta.js
/**
 * Human Delta: relate token reduction to a simple “AI compute savings” score.
 */

import { estimateTokensByModel } from "./tokenEstimate";

/**
 * Percentage token reduction (0–100).
 */
export function reductionPercent(beforeTokens, afterTokens) {
  if (beforeTokens <= 0) return 0;
  const raw = ((beforeTokens - afterTokens) / beforeTokens) * 100;
  return Math.max(0, Math.min(100, Math.round(raw * 10) / 10));
}

/**
 * @param {number} beforeTokens
 * @param {number} afterTokens
 * @returns {{
 *   beforeTokens: number,
 *   afterTokens: number,
 *   efficiencyScore: number,
 *   impactLevel: string
 * }}
 */
export function computeHumanDelta(beforeTokens, afterTokens) {
  const efficiencyScore = reductionPercent(beforeTokens, afterTokens);

  let impactLevel;
  if (efficiencyScore > 60) {
    impactLevel = "LOW IMPACT";
  } else if (efficiencyScore > 30) {
    impactLevel = "MEDIUM IMPACT";
  } else {
    impactLevel = "HIGH IMPACT";
  }

  return {
    beforeTokens,
    afterTokens,
    efficiencyScore,
    impactLevel,
  };
}

/**
 * Convenience: raw + optimized strings → delta for a given tokenizer profile.
 * @param {string} model e.g. "GPT-4" | "Claude" | "LLaMA"
 */
export function humanDeltaFromPrompts(raw, optimized, model = "GPT-4") {
  const beforeTokens = estimateTokensByModel(raw, model);
  const afterTokens = estimateTokensByModel(optimized, model);
  return computeHumanDelta(beforeTokens, afterTokens);
}
